analysis/src/CompletionBackEnd.ml

```diff
@@ -1455,6 +1455,131 @@ let getOpens ~debug ~rawOpens ~package ~env =
+    Some (Completable.Toption (env, payloadTypeExpr))
+  | Tconstr (Path.Pident {name = "bool"}, [], _) -> Some (Tbool env)
+  | Tconstr (path, _, _) -> (
+    match References.digConstructor ~env ~package path with
+    | Some (env, {item = {decl = {type_manifest = Some t1}}}) ->
+      extractType ~env ~package t1
+    | Some (env, {name; item = {decl; kind = Type.Variant constructors}}) ->
+      Some
+        (Tvariant
+           {env; constructors; variantName = name.txt; variantDecl = decl})
+    | _ -> None)
+  | Ttuple expressions -> Some (Tuple (env, expressions))
+  | _ -> None
+
+let completeTypedValue ~env ~full ~prefix ~expandOption =
+  let namesUsed = Hashtbl.create 10 in
+  let rec completeTypedValueInner t ~env ~full ~prefix ~expandOption =
+    let items =
+      match t |> extractType ~env ~package:full.package with
+      | Some (Toption (env, typ)) when expandOption ->
+        typ |> completeTypedValueInner ~env ~full ~prefix ~expandOption:false
+      | Some (Tbool env) ->
+        let items =
+          [
+            Completion.create ~name:"true"
+              ~kind:(Label (t |> Shared.typeToString))
+              ~env;
+            Completion.create ~name:"false"
+              ~kind:(Label (t |> Shared.typeToString))
+              ~env;
+          ]
+        in
+        if prefix = "" then items
+        else
+          items
+          |> List.filter (fun (item : Completion.t) ->
+                 Utils.startsWith item.name prefix)
+      | Some (Tvariant {env; constructors; variantDecl; variantName}) ->
+        let items =
+          constructors
+          |> List.filter_map (fun (constructor : Constructor.t) ->
+                 if
+                   prefix <> ""
+                   && not (Utils.startsWith constructor.cname.txt prefix)
+                 then None
+                 else
+                   Some
+                     (Completion.create
+                        ~name:
+                          (constructor.cname.txt
+                          ^
+                          if constructor.args |> List.length > 0 then
+                            "("
+                            ^ (constructor.args
+                              |> List.map (fun _ -> "_")
+                              |> String.concat ", ")
+                            ^ ")"
+                          else "")
+                        ~kind:
+                          (Constructor
+                             ( constructor,
+                               variantDecl |> Shared.declToString variantName ))
+                        ~env))
+        in
+        items
+      | Some (Toption (env, t)) ->
+        let items =
+          [
+            Completion.create ~name:"None"
+              ~kind:(Label (t |> Shared.typeToString))
+              ~env;
+            Completion.create ~name:"Some(_)"
+              ~kind:(Label (t |> Shared.typeToString))
+              ~env;
```

Comment:
I'm going to improve this a lot soon. Mainly 2 things:
1. `Some(_)` will eventually be a snippet `Some($1:_)`, which means that the cursor is automatically positioned in the option payload (so we can keep completing the option value)
2. If the inner type is "small" enough (bool, variant with few payloads, etc), we can expand those into the list directly, so that instead of `None/Some(_)`, it'd be `None/Some(true)/Some(false)` for bools, etc.